### ISSUE TITLE
Pdf to jpeg conversion doesn't work correctly for some file

### ISSUE THREAD

**Issue by inkhey** [algoo/preview-generator#23]
Currently:
this pdf : 
[programme2018_web.pdf](https://github.com/algoo/preview-generator/files/2387864/programme2018_web.pdf)
 cause issue with jpeg preview. It does not return any exception but stay stuck when preview_generator try to convert pdf to jpeg using "convert_pdf_to_jpeg" method.
Issue seems related to context manager in line 20 of image__wand.py.
" with WImage(file=pdf) as img:" 

if we go deeper, we discover than system got exactly stuck at line 2817 in wand library image.py :
`            r = library.MagickReadImageBlob(self.wand, blob, len(blob))`

Expected: 
Jpeg preview should work on this file too, or at least return an exception instead of being stuck.

**Comment by inkhey** [comment]
after more search, it look like some pdf page generated by pypdf2 are not correctly readable by ghostscript application, that's why jpeg preview from one-page pdf can't be created properly.
You can test this easily : 
[test.pdf](https://github.com/algoo/preview-generator/files/2388666/test.pdf)
this pdf page is created by pypdf, if you tried to transform this pdf into jpeg ghostscript will be stuck.
You can tried this command to verify it  : 
gs -sstdout=%stderr -dQUIET -dSAFER -dBATCH -dNOPAUSE -dNOPROMPT -dMaxBitmap=500000000 -dAlignToPixels=0 -dGridFitTT=2 -sDEVICE=pngalpha -dTextAlphaBits=4 -dGraphicsAlphaBits=4 -r72x72 -sOutputFile=/tmp/output-file.pdf -f/home/XXX/test.pdf

**Comment by buxx** [comment]
@inkhey Maybe open a bug on ghostscript bug tracker ?

**Comment by inkhey** [comment]
@buxx Done https://bugs.ghostscript.com/show_bug.cgi?id=699781

**Comment by PhilippeAccorsi** [comment]
[programme2018_web.pdf](https://github.com/algoo/preview-generator/files/2389307/programme2018_web.pdf)

**Comment by inkhey** [comment]
Problem is solved in last version of ghostscript. 

example with gs 9.20 (debian stable) : 

```
gs test.pdf
GPL Ghostscript 9.20 (2016-09-26)
Copyright (C) 2016 Artifex Software, Inc.  All rights reserved.
This software comes with NO WARRANTY: see the file PUBLIC for details.
Processing pages 1 through 1.
Page 1

```
(it stay stuck there forever)

recent version (gs 9.25 from https://ghostscript.com/download/gsdnld.html) : 

```
 ./gs-925-linux-x86_64 test.pdf
GPL Ghostscript 9.25 (2018-09-13)
Copyright (C) 2018 Artifex Software, Inc.  All rights reserved.
This software comes with NO WARRANTY: see the file PUBLIC for details.
Processing pages 1 through 1.
Page 1
%%BoundingBox: 0 0 420 596
%%HiResBoundingBox: 0.000000 0.000000 419.525987 595.277982
>>showpage, press <return> to continue<<
```

**Comment by buxx** [comment]
ghostscript is available in working version in debian 9 or we have a problem with that ?

**Comment by inkhey** [comment]
@buxx Version from debian 9 is 9.20. bug is solved somewhere between 9.20 and 9.25 (latest). Debian stable actual version can may preview_generator stuck...

**Comment by buxx** [comment]
@inkhey @lebouquetin Okay, so we have a problem ! To solve this issue we have to install newer version of this program (and update install process/scripts).

**Comment by tracim** [comment]
Maybe pypdf2 has to be change for another lib? The original bug is in pypdf2 or in ghostscript?

**Comment by inkhey** [comment]
@tracim Bug is ghostscript related, it's ghostscript who stay stuck anormally with pdf generated by pypdf2.

I'm not sure pdf generated by pypdf2 is fully correct (1.5mo for 1 page of a 1.8mo 10  pages pdf...) but actually appart from this pdf with ghostscript 9.20, i have not see any other issue with generated pdf by this software.

As pypdf2 is not really maintained, change the lib may be useful if we discovered a big issue with it...

**Comment by extremelogic-ph** [comment]
Anyone familiar here to a bug that when a PDF is converted, the image is only black? But it works for some PDF's

**Comment by inkhey** [comment]
Hello @extremelogic-ph 
We do not experience this issue. can you provide us an unworking pdf ?
Depending on the pdf version/features used, the library and software we do use in preview_generator may be buggy.

**Comment by extremelogic-ph** [comment]
[black_when_converted.pdf](https://github.com/algoo/preview-generator/files/4935711/black_when_converted.pdf)

@inkhey  I attached a sample of the PDF that when converted to jpg, it is only a black image.

Tried version 0.13 and the latest 0.14, it is the same.

The way I initiated the conversion is through PreviewManager with just the width and height as parameter. All other PDF's are ok.

**Comment by inkhey** [comment]
> [black_when_converted.pdf](https://github.com/algoo/preview-generator/files/4935711/black_when_converted.pdf)
> 
> @inkhey I attached a sample of the PDF that when converted to jpg, it is only a black image.
> 
> Tried version 0.13 and the latest 0.14, it is the same.
> 
> The way I initiated the conversion is through PreviewManager with just the width and height as parameter. All other PDF's are ok.

Thanks, I can reproduce the issue. Seems like pypdf2 pdf builder can handle properly this case. I do create an issue about using a better pdf library : https://github.com/algoo/preview-generator/issues/192 . If we manage to use Poppler instead, your file should be properly converted as it's the library used by evince pdf viewer and your pdf look great on it.

**Comment by extremelogic-ph** [comment]
> > [black_when_converted.pdf](https://github.com/algoo/preview-generator/files/4935711/black_when_converted.pdf)
> > @inkhey I attached a sample of the PDF that when converted to jpg, it is only a black image.
> > Tried version 0.13 and the latest 0.14, it is the same.
> > The way I initiated the conversion is through PreviewManager with just the width and height as parameter. All other PDF's are ok.
> 
> Thanks, I can reproduce the issue. Seems like pypdf2 pdf builder can handle properly this case. I do create an issue about using a better pdf library : #192 . If we manage to use Poppler instead, your file should be properly converted as it's the library used by evince pdf viewer and your pdf look great on it.

Do we have a schedule on when preview-generator will use the new module to generate the jpeg? If a new version with the new implementation is not yet possible, do you have suggestions to handle this? I am actually using preview-generator as a python web service converter, I find it very useful. But I noticed that not all are converted properly

**Comment by lebouquetin** [comment]
Hello @extremelogic-ph . The upgrade of preview-generator is not yet planned, It may arrive between october and end of the year. Or maybe later. It depends on our clients requests.

A way to fix it by yourself may be to try to upgrad efrom pypdf2 to pypdf4 as it is mentionned by @inkhey in the issue #192.

We are a software development company, we can develop this feature soon for you. Let me know if you are interested in this strategy and I will send you a quote.

**Comment by extremelogic-ph** [comment]
> Hello @extremelogic-ph . The upgrade of preview-generator is not yet planned, It may arrive between october and end of the year. Or maybe later. It depends on our clients requests.
> 
> A way to fix it by yourself may be to try to upgrad efrom pypdf2 to pypdf4 as it is mentionned by @inkhey in the issue #192.
> 
> We are a software development company, we can develop this feature soon for you. Let me know if you are interested in this strategy and I will send you a quote.

I cloned the project in hopes of using the pyPDF4, but I am not yet familiar with almost everything. So I think it will take some time to contribute.

You may send the quote here: extreme.logic@yahoo.com

**Comment by inkhey** [comment]
Theses issues are now solved with the poppler based pdf builder : #275
